### PR TITLE
Use Origin in authorized pallet to resolve dispute for market with Authorized MarketDisputeMechanism.

### LINKED ISSUE #599: [authorized] Allow origins as authorized entity in `authorized`

### PR DESCRIPTION
**Pull request by vivekvpandya**
Fixes #599.

### CODE REVIEW

File: zrml/prediction-markets/src/lib.rs
Comment (Chralt98) [inline zrml/prediction-markets/src/lib.rs:473]: The front-end should be informed in the changelogs for devs file, because multiple extrinsic headers are changed.

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [inline zrml/prediction-markets/src/tests.rs:2456]: Consider using `T::AuthorizedDisputeResolutionUser::get()` here.

**Review by Chralt98 — COMMENTED** [review]
Cool, we have an origin now for authorized, instead of a simple account id.

File: primitives/src/market.rs
Comment (vivekvpandya) [reply primitives/src/market.rs:129]: what happens for old market data which has account_Id in authorized variant of MDM and we try to decode them with this new definition, to me it seems we need migration @maltekliemann please advise :pray:

Comment (maltekliemann) [reply primitives/src/market.rs:129]: Yeah, we should do a storage migration.

The scalar codec encodes instances of enums using one byte to designate which variant is instance is, followed by an encoding of the instances that the variant contains. For example,

```rust
enum OldEnum {
    A,
    B(u8),
}
```

encodes `OldEnum::A` to `0x00` and `OldEnum::B(2)` to `0x0102`.

Furthermore,

> For structures, the values are named, but that is irrelevant for the encoding (names are ignored - only order matters). All containers store elements consecutively. The order of the elements is not fixed, depends on the container, and cannot be relied on at decoding. This implicitly means that decoding some byte-array into a specified structure that enforces an order and then re-encoding it could result in a different byte array than the original that was decoded.

So if you have a struct which contains a field of type `OldEnum` and now replace `OldEnum` with...

```rust
enum NewEnum {
    A,
    B,
}
```

... then its encoding in scale will be a byte shorter. So if you put an `OldMarket` into the decoder of the `NewMarket`, then the scale codec will either reject the byte sequence on the grounds of it being too long, or it will ignore the last byte, giving you a faulty decoding.

I don't know what exactly the scale codec will do in this situation, but I don't intend to find out. However, you can try this with `py-scale-codec` and get a big fat error:

```python
from scalecodec.base import RuntimeConfiguration
from scalecodec.type_registry import load_type_registry_preset

RuntimeConfiguration().update_type_registry(load_type_registry_preset("default"))

my_types = {
    "types": {
        "OldEnum": {
            "type": "enum",
            "type_mapping": [
                ("A", "null"),
                ("B", "u8"),
            ],
        },
        "OldStruct": {
            "type": "struct",
            "type_mapping": [
                ("enum", "OldEnum"),
                ("other", "u8"),
            ],
        },
        "NewEnum": {
            "type": "enum",
            "type_mapping": [
                ("A", "null"),
                ("B", "null"),
            ],
        },
        "NewStruct": {
            "type": "struct",
            "type_mapping": [
                ("enum", "NewEnum"),
                ("other", "u8"),
            ],
        },
    }
}

RuntimeConfiguration().update_type_registry(my_types)

old_struct = RuntimeConfiguration().create_scale_object("OldStruct")
old_raw = old_struct.encode({"enum": {"B": 2}, "other": 3})
print("old_struct: ", old_raw)

new_struct = RuntimeConfiguration().create_scale_object("NewStruct")
new_raw = new_struct.encode({"enum": "B", "other": 3})
print("new_struct: ", new_raw)

new_struct_from_old_raw = RuntimeConfiguration().create_scale_object(
    "NewStruct", old_raw
)
# This raises a scalecodec.exceptions.RemainingScaleBytesNotEmptyException
decoded = new_struct_from_old_raw.decode()
```

**Review by maltekliemann — CHANGES_REQUESTED** [review]
Looks solid. We do, unfortunately, require a storage migration. Also, we should change `misc/types.json` to reflect the changes to `MarketDisputeMechanism`.

File: zrml/authorized/src/benchmarks.rs
Comment (vivekvpandya) [reply zrml/authorized/src/benchmarks.rs:38]: yes I was not aware about other way to solve the issue.

**Comment by sea212** [comment]
> Also, we should change `misc/types.json` to reflect the changes to `MarketDisputeMechanism`.

@yornaath You mentioned that we can derive the `misc/types.json` although "types end up being very verbose". Do you think it is necessary that we manually maintain the (deprecated) `types.json` or can we just derive it?

**Comment by vivekvpandya** [comment]
I think its better to wait until https://github.com/zeitgeistpm/zeitgeist/pull/732/files#diff-7eef70e0e24e29abbdd8ae4c5949b3d7633f1fa7b844e9b0e7ff3227cc344865 these migrations are executed.

**Comment by maltekliemann** [comment]
> I think its better to wait until https://github.com/zeitgeistpm/zeitgeist/pull/732/files#diff-7eef70e0e24e29abbdd8ae4c5949b3d7633f1fa7b844e9b0e7ff3227cc344865 these migrations are executed.

Yeah, sounds good. I've moved the PR to 0.3.7.

File: zrml/market-commons/src/migrations.rs
Comment (maltekliemann) [inline zrml/market-commons/src/migrations.rs:279]:
```suggestion
    }

```

**Review by maltekliemann — CHANGES_REQUESTED** [review]

File: zrml/market-commons/src/migrations.rs
Comment (vivekvpandya) [reply zrml/market-commons/src/migrations.rs:110]: For compiler it can be hard to implement or perhaps algorithm may be very time consuming.

File: zrml/market-commons/src/migrations.rs
Comment (vivekvpandya) [reply zrml/market-commons/src/migrations.rs:131]: should we use same test data as created in `create_test_data_for_market_update()`

Comment (vivekvpandya) [reply zrml/market-commons/src/migrations.rs:131]: if yes then perhaps should we remove test_on_runtime_upgrade() or keep both?

**Comment by Chralt98** [comment]
Starting my review now..

File: zrml/market-commons/src/migrations.rs
Comment (Chralt98) [reply zrml/market-commons/src/migrations.rs:59]: Wow, what a catch Malte! Just a quick reference to the current main:

https://github.com/zeitgeistpm/zeitgeist/blob/5994559f63b1f1b518e22f2987a2eaf8f8d95e20/primitives/src/market.rs#L25-L59

https://github.com/zeitgeistpm/zeitgeist/blob/5994559f63b1f1b518e22f2987a2eaf8f8d95e20/primitives/src/market.rs#L129-L135

File: zrml/market-commons/src/migrations.rs
Comment (Chralt98) [reply zrml/market-commons/src/migrations.rs:87]: Rust derives the type with the first push to the vector. The explicit declaration by the coder is useful for a better understanding sometimes.

File: zrml/market-commons/src/migrations.rs
Comment (Chralt98) [reply zrml/market-commons/src/migrations.rs:110]: It doesn't work, because the new market has pedantically a different type than the old market.

File: zrml/market-commons/src/migrations.rs
Comment (Chralt98) [reply zrml/market-commons/src/migrations.rs:131]: I mean all migrations are transitory. So code duplication is no problem on my side here (keep both). @maltekliemann that sounds like a good approach for validation.

> should we use same test data as created in create_test_data_for_market_update()

`try-runtime` can test based on the real blockchain storage. So you basically query and save all old data which is currently on the blockchain and check that all fields are still present (and equal) after the upgrade. The type for old markets is [`LegacyMarket`](https://github.com/zeitgeistpm/zeitgeist/blob/5994559f63b1f1b518e22f2987a2eaf8f8d95e20/primitives/src/market.rs#L25-L59).

Comment (maltekliemann) [reply zrml/market-commons/src/migrations.rs:131]: Yeah, we should keep both, because they're actually testing slightly different behaviors.

File: zrml/market-commons/src/migrations.rs
Comment (maltekliemann) [reply zrml/market-commons/src/migrations.rs:110]: > For compiler it can be hard to implement or perhaps algorithm may be very time consuming.

I guess. Still, would be cool if we could do this. ;)

> It doesn't work, because the new market has pedantically a different type than the old market.

Yeah, but the remaining fields all have the same type. That's why I had hoped this would work.

File: zrml/market-commons/src/migrations.rs
Comment (maltekliemann) [inline zrml/market-commons/src/migrations.rs:155]: It's not super important, but I would not iterate over `storage_iter::<MarketOf<T>>(MARKET_COMMONS, MARKETS)` and then use `get_temp_storage` in the loop. The behavior you're testing here is: "After the migration, for every `MarketOf<T>` contained in `Markets`, the corresponding legacy market (in the temporary storage) has the same fields, except for the changed dispute mechanism."

The problem with this is that there are unwanted behaviors which pass this test. For example, if the migration just deletes all markets, it passes this test.

Instead, you want to test: "After the migration, for every legacy market (in the temporary storage), the corresponding new market has the same fields, except for the changed dispute mechanism."

**Comment by maltekliemann** [comment]
Starting my review...

File: zrml/market-commons/src/migrations.rs
Comment (maltekliemann) [inline zrml/market-commons/src/migrations.rs:156]: Could you add a test that verifies that the correct pallet version has been set?

File: zrml/market-commons/src/migrations.rs
Comment (maltekliemann) [inline zrml/market-commons/src/migrations.rs:220]: The difference between these two test cases is that one is categorical, while one is scalar. Woudn't it me more useful to have one market that has `Authorized` as MDM and one that has, say, `Court`, so that we can check if `Court` is correctly changed. I mean, the migration doesn't care if the market is categorical or scalar, but it does care very much what the dispute mechanism is.

**Review by maltekliemann — COMMENTED** [review]
Waiting until the merge conflicts are done with my final review.

File: zrml/market-commons/src/migrations.rs
Comment (vivekvpandya) [reply zrml/market-commons/src/migrations.rs]: Try to address this by storing the count of legacy markets and then comparing it against number of market found in post_upgrade()

File: zrml/market-commons/src/migrations.rs
Comment (Chralt98) [inline zrml/market-commons/src/migrations.rs:161]:
```suggestion
            assert_eq!(updated_market.resolved_outcome, legacy_market.resolved_outcome);
            let dispute_mechanism = match legacy_market.dispute_mechanism {
                LegacyMarketDisputeMechanism::Authorized(_) => MarketDisputeMechanism::Authorized,
                LegacyMarketDisputeMechanism::Court => MarketDisputeMechanism::Court,
                LegacyMarketDisputeMechanism::SimpleDisputes => {
                    MarketDisputeMechanism::SimpleDisputes
                }
            };
            assert_eq!(updated_market.dispute_mechanism, dispute_mechanism)
```

**Review by Chralt98 — DISMISSED** [review]
Saved storage migration. The ability to save data in `pre_upgrade` and read it in `post_upgrade` is useful.

**Review by Chralt98 — APPROVED** [review]

**Review by maltekliemann — CHANGES_REQUESTED** [review]

File: zrml/prediction-markets/src/migrations.rs
Comment (vivekvpandya) [reply zrml/prediction-markets/src/migrations.rs:226]: Oh sorry completely missed it.

File: runtime/common/src/lib.rs
Comment (vivekvpandya) [reply runtime/common/src/lib.rs:63]: Thanks for thinking of this!

**Review by Chralt98 — APPROVED** [review]

File: zrml/prediction-markets/src/migrations.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/migrations.rs:226]: No problem, thanks for fixing.

File: zrml/market-commons/src/migrations.rs
Comment (maltekliemann) [inline zrml/market-commons/src/migrations.rs:19]:
```suggestion
use alloc::vec::Vec;
```

**Review by maltekliemann — APPROVED** [review]
:thumbsup:

**Review by maltekliemann — APPROVED** [review]

**Pull request merged**
